Render MathML unknown elements as mrow.

eerii opened the pull request:
#### 752e87ed0dc7dcad7a0787defb8cb6ca175bb985
<pre>
Render MathML unknown elements as mrow.
<a href="https://bugs.webkit.org/show_bug.cgi?id=290625">https://bugs.webkit.org/show_bug.cgi?id=290625</a>

Reviewed by Frédéric Wang.

The MathML Core spec (<a href="https://w3c.github.io/mathml-core/#new-display-math-value)">https://w3c.github.io/mathml-core/#new-display-math-value)</a>
says that &quot;Unknown MathML elements behave the same as the mrow element&quot;.

* LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/mrow/inferred-mrow-stretchy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/mrow/spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/operators/embellished-operator-001-expected.txt:
* LayoutTests/mathml/mathml-unknown.xhtml: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/presentation-markup/mrow/inferred-mrow-baseline-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/presentation-markup/mrow/mrow-preferred-width-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/presentation-markup/mrow/inferred-mrow-baseline-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/presentation-markup/mrow/mrow-preferred-width-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/mathml/presentation-markup/mrow/inferred-mrow-baseline-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/mathml/presentation-markup/mrow/mrow-preferred-width-expected.txt:
* Source/WebCore/mathml/MathMLPresentationElement.h:
* Source/WebCore/mathml/MathMLRowElement.cpp:
(WebCore::MathMLRowElement::createElementRenderer):
* Source/WebCore/mathml/MathMLUnknownElement.h:

Canonical link: <a href="https://commits.webkit.org/300580@main">https://commits.webkit.org/300580@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/00ad656a0cd1bc0a0ea04cb8eef1a76b2c7b568e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123097 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42811 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33508 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129756 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75207 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124974 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43535 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51406 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93570 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/62091 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126048 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34694 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110169 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74205 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33664 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28325 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73271 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104406 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28551 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132485 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50047 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38098 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-auto-text-fragment.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102070 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50423 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106390 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101921 "Found 3 new API test failures: TestWTF:CompletionHandlerDeathTest.MainThreadHandlerOnThreadAssertsDeathTest, WebKitGTK/TestInspector:/webkit/WebKitWebInspector/manual-attach-detach, TestWTF:CompletionHandlerDeathTest.ConstructionThreadHandlerOnThreadAssertsDeathTest (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25912 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47297 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25493 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/46820 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49902 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55663 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49370 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52722 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51051 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e31f99aff5ca3a93432ec63c6ce5f08540beaee9. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51189

Ahmad-S792 commented:
https://w3c.github.io/mathml-core/#new-display-math-value

"Unknown MathML elements behave the same as the mrow element."

Can we add it in our commit message? Plus you need to remove 'No new tests (OOPS!).' else automation bot will not merge it.

fred-wang commented on LayoutTests/TestExpectations:
```diff
@@ -1891,6 +1891,13 @@ imported/w3c/web-platform-tests/mathml/presentation-markup/operators/mo-form-dyn
 webkit.org/b/201356 mathml/presentation/stretchy-depth-height-symmetric.html [ Skip ]
 
+# These internal MathML tests are outdated and do not match MathML Core.
+mathml/mathml-unknown.xhtml [ Failure ]
```
I think we can just remove this test, as it's clearly wrong with this patch.

fred-wang commented on LayoutTests/TestExpectations:
```diff
@@ -1891,6 +1891,13 @@ imported/w3c/web-platform-tests/mathml/presentation-markup/operators/mo-form-dyn
+mathml/presentation/semantics-2.html [ ImageOnlyFailure ]
+mathml/presentation/semantics-3.html [ ImageOnlyFailure ]
+mathml/presentation/semantics-4.html [ ImageOnlyFailure ]
+
```
I think we will eventually remove these tests when we align semantics/maction with MathML Core, but I don't understand why they are failing with your change. Can you explain why? Maybe there is a way to tweak them so they still pass for now.

fred-wang commented on Source/WebCore/mathml/MathMLUnknownElement.h:
```diff
@@ -27,11 +27,11 @@
 namespace WebCore {
 
-class MathMLUnknownElement final : public MathMLElement {
+class MathMLUnknownElement final : public MathMLRowElement {
```
I wonder if this MathMLUnknownElement class is still needed... not sure where we actually use it.

eerii replied:
It's a good point. However, in the end if we are overriding `isPresentationMathML`, let's keep it until we no longer care about that distinction.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/dc2c15942c1fb1b256094eea06ee73c83da60e50. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51189

fred-wang commented on Source/WebCore/mathml/MathMLUnknownElement.h:
```diff
@@ -42,11 +42,11 @@ class MathMLUnknownElement final : public MathMLElement {
     }
 
-    bool rendererIsNeeded(const RenderStyle&) final { return false; }
+    bool isPresentationMathML() const override { return false; }
```
nit: this can be s/override/final/

fred-wang reviewed (approved):
Thanks LGTM.

eerii committed:
Render MathML unknown elements as mrow.
https://bugs.webkit.org/show_bug.cgi?id=290625

Reviewed by Frédéric Wang.

The MathML Core spec (https://w3c.github.io/mathml-core/#new-display-math-value)
says that "Unknown MathML elements behave the same as the mrow element".

* LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/mrow/inferred-mrow-stretchy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/mrow/spacing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/operators/embellished-operator-001-expected.txt:
* LayoutTests/mathml/mathml-unknown.xhtml: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/presentation-markup/mrow/inferred-mrow-baseline-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/presentation-markup/mrow/mrow-preferred-width-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/presentation-markup/mrow/inferred-mrow-baseline-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/presentation-markup/mrow/mrow-preferred-width-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/mathml/presentation-markup/mrow/inferred-mrow-baseline-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/mathml/presentation-markup/mrow/mrow-preferred-width-expected.txt:
* Source/WebCore/mathml/MathMLPresentationElement.h:
* Source/WebCore/mathml/MathMLRowElement.cpp:
(WebCore::MathMLRowElement::createElementRenderer):
* Source/WebCore/mathml/MathMLUnknownElement.h:

Canonical link: https://commits.webkit.org/300580@main
